Applications/Spire/Include/Spire/Ui/MenuButton.hpp

```diff
@@ -0,0 +1,36 @@
+#ifndef SPIRE_MENU_BUTTON_HPP
+#define SPIRE_MENU_BUTTON_HPP
+#include <QTimer>
+#include <QWidget>
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  /** Represents a button that can open a menu */
+  class MenuButton : public QWidget {
+    public:
+
+    /**
+     * Constructs a MenuButton.
+     * @param body The component that opens a menu.
+     * @param menu The menu that is revealed when the body is clicked.
+     * @param parent The parent widget.
+     */
+    MenuButton(QWidget& body, OverlayPanel& menu, QWidget* parent = nullptr);
+
+    protected:
+      bool eventFilter(QObject* watched, QEvent* event) override;
+      void keyPressEvent(QKeyEvent* event) override;
+      void keyReleaseEvent(QKeyEvent* event) override;
+      void mousePressEvent(QMouseEvent* event) override;
+      void mouseReleaseEvent(QMouseEvent* event) override;
+
+    private:
+      OverlayPanel* m_menu;
+      QTimer m_timer;
+      bool m_is_mouse_down_on_button_with_menu;
```

Comment:
Are these ```bool``` variables independent of each other?

Reply:
Yes. m_is_mouse_down_on_button_with_menu indicates that the mouse pressed down on the button when the menu is visible, while m_is_mouse_down_on_button_without_menu indicates that the mouse pressed down on the button when the menu is invisible. They are used for different purpose.

Reply:
I think I can remove one of them.

Reply:
Another option is if they are not independent to use an enum to specify a set of states.